Adding resource requests for control-plane & agents-plane components (#109)
litmuschaos/litmus-docs: docs/pod-network-latency.md

---
id: pod-network-latency
title: Pod Network Latency Experiment Detail
sidebar_label: Pod Network Latency
---
------

## Experiment Metadata

| Type      | Description              | Tested K8s Platform                                               |
| ----------| ------------------------ | ------------------------------------------------------------------|
| Generic   | Inject Network Latency Into Application Pod | Any|

## Prerequisites

- Ensure that the Litmus Chaos Operator is running

> - Experimenting Cluster should be non-minikube cluster

Reviewer: Can we put this as a note in the pre-requisites that minikube is not supported & also change the tested K8s platform to the actual ones (you can refer to the pod-delete/container-kill cases). Right now it is a bit contradictory (Any/Minikube not supported) We can discuss addition of a `supported platforms` section to encourage the community.

> - Application subjected to chaos must have tc network traffic shaping tool installed

Reviewer: Application container

- Ensure that the `pod-network-latency` experiment resource is available in the cluster. If not, install from [here](https://hub.litmuschaos.io/charts/generic/experiments/pod-network-latency)

> 

Reviewer: Lets add a note to say that the experiment is supported only on Docker Runtime. We do not support containerd/CRIO runtimes yet for network tests. There is a way for us to directly invoke `tc`, but these utils aren't added yet.

Reviewer 2: Let's also add that on minikube, the container must be priviledged, otherwise it has no permissions to mess with tc netem ...

Reviewer 2: So same as with the loss experiment: minikube works as soon as my pr gets merged by them, but the container must be privileged in order to execute the tc netem command.

Reviewer: @ashish-uiet can you take care of the above?

Reviewer 3: waiting for code merge @LaumiH .
as of now checked in minikube following conf but not injecting chaos -
image - ksatchit/ansible-runner:ci
LIB_IMAGE as pumba 0.6.5

Reviewer: With latest changes to the network chaos experiments, an iproute2 container is necessary to inject the chaos (was done to facilitate the cases where app containers' base images do not have tc installed)

Reviewer: Probably we could add the note about minikube support in the README on we have more info/test results that @LaumiH is helping us with..

Reviewer 2: Sorry, I was on sick leave yesterday. I tested it with the newest minikube, it works *without* privileged containers now, with pumba 0.6.5.

test pod:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: busy3
  name: busy3
spec:
  replicas: 1
  selector:
    matchLabels:
      app: busy3
  template:
    metadata:
       labels:
        app: busy3
    spec:
      containers:
        - image: debian
          name: busy-3
          command: [ "/bin/bash", "-c", "sleep 10000;exit 0" ]
```
chaosengine:
```
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: engine
  namespace: default
spec:
  jobCleanUpPolicy: retain
  monitoring: false
  appinfo: 
    appns: default
    applabel: 'app=busy3'
    appkind: deployment 
  chaosServiceAccount: 'busy'
  experiments:
    - name: pod-network-loss
      spec:
        components:
        - name: TARGET_CONTAINER
          value: 'busy-3'
        - name: NETWORK_INTERFACE
          value: 'eth0'
        - name: TOTAL_CHAOS_DURATION
          value: '40000' #ms
        - name: LIB
          value: 'pumba'
        - name: LIB_IMAGE
          value: 'gaiaadm/pumba:latest'
```

## Entry Criteria

- Application pods are healthy before chaos injection

## Exit Criteria

- Application pods are healthy post chaos injection

## Details

- Pod-network-latency contains chaos to disrupt network connectivity of kubernetes pods.

> - The test involved setting up a ping to general/public IPs from inside the pod & also setting up a ping to the pod IP itself from a cluster node.

Reviewer: Not sure if the test needs to find a mention. The users of the experiment might have their own ways to detect the chaos injected. However, if you really feel like adding this, lets add a section on `example test` where you describe how it can be done.

- The application pod should be healthy once chaos is stopped. Service-requests should be served despite chaos.

> - Causes flaky access to application replica by injecting network delay using pumba.

Reviewer: Probably add some more description from these points: 

- Injects latency on the specified container by starting a traffic control (`tc`) process with netem rules to add egress delays
- Latency is injected via pumba library with command `pumba netem delay` by passing the relevant network interface, latency, chaos duration and regex filter for container name
- Pumba is run as a daemonset on all nodes in dry-run mode to begin with; the latency command is issued during experiment execution via kubectl exec
Can test the application's resilience to lossy/flaky network

## Steps to Execute the Chaos Experiment

- This Chaos Experiment can be triggered by creating a ChaosEngine resource on the cluster. To understand the values to provide in a ChaosEngine specification, refer [Getting Started](getstarted.md/#prepare-chaosengine)

- Follow the steps in the sections below to prepare the ChaosEngine & execute the experiment.

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`
- Override the experiment tunables if desired

#### Supported Experiment Tunables

| Variables             | Description                                                  | Type      | Notes                                                      |
| ----------------------| ------------------------------------------------------------ |-----------|------------------------------------------------------------|

> | TOTAL_CHAOS_DURATION  | The time duration for chaos insertion (seconds)              | Mandatory  | 60000                                            |

Reviewer: TOTAL_CHAOS_DURATION & NETWORK_LATENCY have default values. Hence Optional. The target container, network interface are generally unknown/instance-specific entities. These mandatory ENVs can probably be grouped upfront followed by the Optional ones

| NETWORK_LATENCY        | The latency/delay in milliseconds                           | Mandatory  | 60000
| LIB                   | The chaos lib used to inject the chaos eg. Pumba             | Optional  |  |
| NETWORK_INTERFACE     | Name of ethernet interface considered for shaping traffic                                | Mandatory  |   |
| TARGET_CONTAINER     | Name of container which is subjected to network latency      | Mandatory  |   |
| CHAOSENGINE     | ChaosEngine CR name associated with the experiment instance      | Optional  |   |
| CHAOS_SERVICE_ACCOUNT     | Service account used by the pumba daemonset Optional      | Optional  |   |

#### Sample ChaosEngine Manifest

```yaml
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:

>   name: engine-nginx

Reviewer: Lets probably name it nginx-chaos OR nginx-network-chaos

  namespace: default
spec:
  jobCleanUpPolicy: retain
  monitoring: false
  appinfo: 
    appns: default
    # FYI, To see app label, apply kubectl get pods --show-labels
    applabel: "app=nginx-app"
    appkind: deployment
  chaosServiceAccount: nginx 
  experiments:
    - name: pod-network-latency
      spec:
        components:
        - name: ANSIBLE_STDOUT_CALLBACK
          value: default
        - name: TARGET_CONTAINER

>           value: nginx-deploy-container

Reviewer: lets take a look at other experiments, such as the container-kill where an nginx is being used as the sample application. Lets use the same application specs to keep the doc consistent/homogeneous. Say, someone installs the nginx app as recommended in docs  - they should be able to pick the sample chaosEngines to just work!

Reviewer 3: okay

        - name: NETWORK_INTERFACE
          value: eth0
        - name: LIB_IMAGE
          value: gaiaadm/pumba:0.4.8
        - name: NETWORK_LATENCY
          value: "60000"
        - name: TOTAL_CHAOS_DURATION
          value: "60000"
        - name: LIB
          value: pumba
```
### Create the ChaosEngine Resource

- Create the ChaosEngine manifest prepared in the previous step to trigger the Chaos.

  `kubectl apply -f chaosengine.yml`

### Check Chaos Experiment Result

- Check whether the application is resilient to the Pod Network Latency, once the experiment (job) is completed. The ChaosResult resource name is derived like this: `<ChaosEngine-Name>-<ChaosExperiment-Name>`.

  `kubectl describe chaosresult <ChaosEngine-Name>-<ChaosExperiment-Name> -n <application-namespace>`

> ### Test Chaos progress

Reviewer: Watch Chaos Progress

Reviewer: The section is expected before the Check Chaos Experiment Result

> - During Chaos progress interval View Pod Network Latency inside targeted container , By setting Up Traffic control (tc) tool in targeted container .Check Ping Simulate Internet connections.

Reviewer: ```suggestion
- View network latency by setting up a ping on the affected pod from the cluster nodes 
```

| LIB                   | The chaos lib used to inject the chaos eg. Pumba             | Optional  |  |
| NETWORK_INTERFACE     | Name of ethernet interface considered for shaping traffic                                | Mandatory  |   |
| TARGET_CONTAINER     | Name of container which is subjected to network latency      | Mandatory  |   |
| CHAOSENGINE     | ChaosEngine CR name associated with the experiment instance      | Optional  |   |
| CHAOS_SERVICE_ACCOUNT     | Service account used by the pumba daemonset Optional      | Optional  |   |

#### Sample ChaosEngine Manifest

```yaml
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name:  nginx-network-chaos
  namespace: default
spec:
  jobCleanUpPolicy: retain
  monitoring: false
  appinfo: 
    appns: default
    # FYI, To see app label, apply kubectl get pods --show-labels
    applabel: "app=nginx-app"
    appkind: deployment
  chaosServiceAccount: nginx
  experiments:
    - name: pod-network-latency
      spec:
        components:
        - name: ANSIBLE_STDOUT_CALLBACK
          value: default
        - name: TARGET_CONTAINER

>           value: "nginx-deploy-container" #Container name where chaos has to be injected

Reviewer: nginx

---
id: pod-network-latency

> title: Pod Network Latency Experiment Detail

Reviewer: Details

sidebar_label: Pod Network Latency
---
------

## Experiment Metadata

| Type      | Description              | Tested K8s Platform                                               |
| ----------| ------------------------ | ------------------------------------------------------------------|
| Generic   | Inject Network Latency Into Application Pod | GKE, Konvoy(AWS), Packet(Kubeadm), Minikube, OpenShift(Baremetal)|

## Prerequisites
- Ensure that the Litmus Chaos Operator is running

> - Application subjected to chaos must have tc network traffic shaping tool installed

Reviewer: With our new enhancements based  on 0.6.5, we longer depend on tc

Reviewer 3: removed 16th line

- This Chaos Experiment can be triggered by creating a ChaosEngine resource on the cluster. To understand the values to provide in a ChaosEngine specification, refer [Getting Started](getstarted.md/#prepare-chaosengine)

- Follow the steps in the sections below to prepare the ChaosEngine & execute the experiment.

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`
- Override the experiment tunables if desired

#### Supported Experiment Tunables

| Variables             | Description                                                  | Type      | Notes                                                      |
| ----------------------| ------------------------------------------------------------ |-----------|------------------------------------------------------------|
| TOTAL_CHAOS_DURATION  | The time duration for chaos insertion (seconds)              | Optional  | 60000                                            |
| NETWORK_LATENCY        | The latency/delay in milliseconds                           | Optional  | 60000
| LIB                   | The chaos lib used to inject the chaos eg. Pumba             | Optional  |  |

> | NETWORK_INTERFACE     | Name of ethernet interface considered for shaping traffic                                | Instance-Specific (Optional)  |   |

Reviewer: Target Container and Network Interface can be made mandatory.

| TARGET_CONTAINER     | Name of container which is subjected to network latency      | Instance-Specific (Optional)  |   |
| CHAOSENGINE     | ChaosEngine CR name associated with the experiment instance      | Optional  |   |
| CHAOS_SERVICE_ACCOUNT     | Service account used by the pumba daemonset Optional      | Optional  |   |

> #### Sample ChaosEngine Manifest

Reviewer: Like discussed, please change this to nginx

Reviewer 3: done

kind: ChaosEngine
metadata:
  name:  redis-network-chaos
  namespace: default
spec:
  jobCleanUpPolicy: retain
  monitoring: false
  appinfo: 
    appns: default
    # FYI, To see app label, apply kubectl get pods --show-labels
    applabel: "app=redis-app"
    appkind: deployment
  chaosServiceAccount: redis
  experiments:
    - name: pod-network-latency
      spec:
        components:
        - name: ANSIBLE_STDOUT_CALLBACK
          value: default
        - name: TARGET_CONTAINER
          value: redis-deploy-container #Container name where chaos has to be injected
        - name: NETWORK_INTERFACE
          value: eth0                   #Network interface inside target container
        - name: LIB_IMAGE
          value: gaiaadm/pumba:0.6.5
        - name: NETWORK_LATENCY
          value: "2000"
        - name: TOTAL_CHAOS_DURATION
          value: "60000"
        - name: LIB
          value: pumba
```
### Create the ChaosEngine Resource

- Create the ChaosEngine manifest prepared in the previous step to trigger the Chaos.

  `kubectl apply -f chaosengine.yml`

### Watch Chaos progress

- View network latency by setting up a ping on the affected pod from the cluster nodes 

>     `ping http_address`

Reviewer: can we change this to`ping <pod_ip_address`

Reviewer 3: done
